# Day 13: Transparent Origami

## Part one

You reach another volcanically active part of the cave. It would be nice if you could do some kind of thermal imaging so you could tell ahead of time which caves are too hot to safely enter.

Fortunately, the submarine seems to be equipped with a thermal camera! When you activate it, you are greeted with:

```
Congratulations on your purchase! To activate this infrared thermal imaging
camera system, please enter the code found on page 1 of the manual.
```

Apparently, the Elves have never used this feature. To your surprise, you manage to find the manual; as you go to open it, page 1 falls out. It's a large sheet of *transparent paper*! The transparent paper is marked with random dots and includes instructions on how to fold it up (your puzzle input). For example:

```
6,10
0,14
9,10
0,3
10,4
4,11
6,0
6,12
4,1
0,13
10,12
3,4
3,0
8,4
1,10
2,14
8,10
9,0

fold along y=7
fold along x=5
```

The first section is a list of dots on the transparent paper. `0,0` represents the top-left coordinate. The first value, `x`, increases to the right. The second value, `y`, increases downward. So, the coordinate 3,0 is to the right of `0,0`, and the coordinate `0,7` is below `0,0`. The coordinates in this example form the following pattern, where `#` is a dot on the paper and `.` is an empty, unmarked position:

```
...#..#..#.
....#......
...........
#..........
...#....#.#
...........
...........
...........
...........
...........
.#....#.##.
....#......
......#...#
#..........
#.#........
```

Then, there is a list of fold instructions. Each instruction indicates a line on the transparent paper and wants you to fold the paper up (for horizontal y=... lines) or left (for vertical x=... lines). In this example, the first fold instruction is fold along y=7, which designates the line formed by all of the positions where y is 7 (marked here with -):

```
...#..#..#.
....#......
...........
#..........
...#....#.#
...........
...........
-----------
...........
...........
.#....#.##.
....#......
......#...#
#..........
#.#........
```

Because this is a horizontal line, fold the bottom half up. Some of the dots might end up overlapping after the fold is complete, but dots will never appear exactly on a fold line. The result of doing this fold looks like this:

```
#.##..#..#.
#...#......
......#...#
#...#......
.#.#..#.###
...........
...........
```

Now, only 17 dots are visible.

Notice, for example, the two dots in the bottom left corner before the transparent paper is folded; after the fold is complete, those dots appear in the top left corner (at 0,0 and 0,1). Because the paper is transparent, the dot just below them in the result (at 0,3) remains visible, as it can be seen through the transparent paper.

Also notice that some dots can end up **overlapping**; in this case, the dots merge together and become a single dot.

The second fold instruction is fold along x=5, which indicates this line:

```
#.##.|#..#.
#...#|.....
.....|#...#
#...#|.....
.#.#.|#.###
.....|.....
.....|.....
```

Because this is a vertical line, fold **left**:

```
#####
#...#
#...#
#...#
#####
.....
.....
```

The instructions made a square!

The transparent paper is pretty big, so for now, focus on just completing the first fold. After the first fold in the example above, 17 dots are visible - dots that end up overlapping after the fold is completed count as a single dot.

**How many dots are visible after completing just the first fold instruction on your transparent paper?**

In [1]:
import re

def load_input(filepath='./data/input_test.txt'):
    dot_map = set()
    instructions = []
    with open('%s' % (filepath, ), 'r') as input_f:
        line_number = 0
        input_lines = input_f.read().splitlines()
        
        # Extract dot map
        while True:
            input_line = input_lines[line_number]
            line_number += 1
            if not input_line:
                break
            x, y = input_line.split(',')
            dot_map.add((int(x), int(y)))
        
        # Extract instructions
        instruction_line_number = line_number
        for line_number in range(instruction_line_number, len(input_lines)):
            input_line = input_lines[line_number]
            line_number += 1
            if not input_line:
                break
            
            reg_match = re.match('^fold along (x|y)=([0-9]+)$', input_line)
            direction = reg_match[1]
            value = reg_match[2]
            instruction = { 'direction': direction, 'value': int(value) }
            instructions.append(instruction)
    
    return dot_map, instructions

dot_map, instructions = load_input()
print('dot_map: %s' % (dot_map, ))
print('instructions: %s' % (instructions, ))
len(dot_map)

dot_map: {(9, 10), (6, 12), (2, 14), (9, 0), (8, 4), (3, 4), (10, 4), (0, 13), (0, 3), (8, 10), (3, 0), (6, 10), (10, 12), (6, 0), (1, 10), (4, 11), (0, 14), (4, 1)}
instructions: [{'direction': 'y', 'value': 7}, {'direction': 'x', 'value': 5}]


18

In [2]:
def execute_instruction(instruction, dot_map):
    new_dot_map = set()
    if instruction['direction'] == 'x':
        element_number = 0
    else:
        element_number = 1
    
    for element in dot_map:
        new_position = [element[0], element[1]]
        if element[element_number] >= instruction['value']:
            new_position[element_number] = instruction['value'] - (element[element_number] - instruction['value'])
        
        new_dot_map.add(tuple(new_position))
    
    return new_dot_map

In [3]:
dot_map, instructions = load_input()
new_dot_map = execute_instruction({'direction': 'y', 'value': 7}, dot_map)
print('new_dot_map: %s' % (new_dot_map, ))
print('len new_dot_map: %s' % (len(new_dot_map), ))
new_dot_map

new_dot_map: {(0, 1), (9, 0), (6, 2), (8, 4), (3, 4), (4, 3), (0, 0), (10, 4), (0, 3), (2, 0), (6, 4), (1, 4), (3, 0), (6, 0), (4, 1), (10, 2), (9, 4)}
len new_dot_map: 17


{(0, 0),
 (0, 1),
 (0, 3),
 (1, 4),
 (2, 0),
 (3, 0),
 (3, 4),
 (4, 1),
 (4, 3),
 (6, 0),
 (6, 2),
 (6, 4),
 (8, 4),
 (9, 0),
 (9, 4),
 (10, 2),
 (10, 4)}

In [4]:
dot_map, instructions = load_input('./data/input.txt')
new_dot_map = execute_instruction(instructions[0], dot_map)
print('len new_dot_map: %s' % (len(new_dot_map), ))

len new_dot_map: 810


Your puzzle answer was `810`.

## Part two

Finish folding the transparent paper according to the instructions. The manual says the code is always **eight capital letters**.

**What code do you use to activate the infrared thermal imaging camera system?**

In [5]:
dot_map, instructions = load_input('./data/input.txt')
for instruction in instructions:
    print('instruction: %s' % (instruction, ))
    dot_map = execute_instruction(instruction, dot_map)
print('len dot_map: %s' % (len(dot_map), ))

instruction: {'direction': 'x', 'value': 655}
instruction: {'direction': 'y', 'value': 447}
instruction: {'direction': 'x', 'value': 327}
instruction: {'direction': 'y', 'value': 223}
instruction: {'direction': 'x', 'value': 163}
instruction: {'direction': 'y', 'value': 111}
instruction: {'direction': 'x', 'value': 81}
instruction: {'direction': 'y', 'value': 55}
instruction: {'direction': 'x', 'value': 40}
instruction: {'direction': 'y', 'value': 27}
instruction: {'direction': 'y', 'value': 13}
instruction: {'direction': 'y', 'value': 6}
len dot_map: 103


In [6]:
def print_map(dot_map, size_x=40, size_y=6):
    for y in range(0, size_y):
        line = ''
        for x in range(0, size_x):
            if (x, y) in dot_map:
                line += '▓'
            else:
                line += '░'
        print(line)

print_map(dot_map)            

▓░░▓░▓░░░░▓▓▓░░▓░░▓░▓▓▓░░░▓▓░░▓▓▓▓░▓▓▓░░
▓░░▓░▓░░░░▓░░▓░▓░░▓░▓░░▓░▓░░▓░▓░░░░▓░░▓░
▓▓▓▓░▓░░░░▓▓▓░░▓░░▓░▓▓▓░░▓░░░░▓▓▓░░▓░░▓░
▓░░▓░▓░░░░▓░░▓░▓░░▓░▓░░▓░▓░▓▓░▓░░░░▓▓▓░░
▓░░▓░▓░░░░▓░░▓░▓░░▓░▓░░▓░▓░░▓░▓░░░░▓░▓░░
▓░░▓░▓▓▓▓░▓▓▓░░░▓▓░░▓▓▓░░░▓▓▓░▓░░░░▓░░▓░


Your puzzle answer was `HLBUBGFR`.